packages/hardhat-core/src/internal/hardhat-network/provider/TxPool.ts

```diff
@@ -473,10 +531,132 @@ export class TxPool {
     this._state = this._state.set("blockGasLimit", bnToHex(newLimit));
   }
 
+  private _setMinGasPrice(newMin: BN) {
+    this._state = this._state.set("minGasPrice", bnToHex(newMin));
+  }
+
   private _deleteTransactionByHash(hash: Buffer) {
     this._state = this._state.set(
       "hashToTransaction",
       this._getTransactionsByHash().delete(bufferToHex(hash))
     );
   }
+
+  private _isTxValid(
+    tx: OrderedTransaction,
+    txNonce: BN,
+    senderNonce: BN,
+    senderBalance: BN
+  ): boolean {
+    const txGasLimit = new BN(tx.data.gasLimit);
+    const txGasPrice = new BN(tx.data.gasPrice);
+
+    return (
+      txGasLimit.lte(this.getBlockGasLimit()) &&
+      txNonce.gte(senderNonce) &&
+      tx.data.getUpfrontCost().lte(senderBalance) &&
+      txGasPrice.gte(this.getMinGasPrice())
+    );
+  }
+
+  /**
+   * Returns the next available nonce for an address, ignoring its
+   * pending transactions.
+   */
+  private async _getAccountNonce(accountAddress: Buffer): Promise<BN> {
+    const account = await this._stateManager.getAccount(accountAddress);
+    return new BN(account.nonce);
+  }
+
+  /**
+   * Checks if some pending tx with the same nonce as `newTx` exists.
+   * If it exists, it replaces it with `newTx` and returns true.
+   * Otherwise returns false.
+   */
+  private _replacePendingTx(
+    accountAddress: string,
+    newTx: OrderedTransaction
+  ): boolean {
+    const pendingTxs = this._getPendingForAddress(accountAddress);
+    const newPendingTxs = this._replaceTx(pendingTxs, newTx);
+
+    if (newPendingTxs !== undefined) {
+      this._setPendingForAddress(accountAddress, newPendingTxs);
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Checks if some queued tx with the same nonce as `newTx` exists.
+   * If it exists, it replaces it with `newTx` and returns true.
+   * Otherwise returns false.
+   */
+  private _replaceQueuedTx(
+    accountAddress: string,
+    newTx: OrderedTransaction
+  ): boolean {
+    const queuedTxs = this._getQueuedForAddress(accountAddress);
+    const newQueuedTxs = this._replaceTx(queuedTxs, newTx);
+
+    if (newQueuedTxs !== undefined) {
+      this._setQueuedForAddress(accountAddress, newQueuedTxs);
+      return true;
+    }
+
+    return false;
+  }
+
+  private _replaceTx(
```

Comment:
This method has logic that is common to both `_replacePendingTx` and `_replaceQueuedTx`